### PR TITLE
BN: Charity - Testing

### PR DESCRIPTION
**Pull request by Sphyxis**
Charity BN - Testing

Presenting the Charity BN for testing.  A gang inspired challenging BN.
Github is hosted here:
https://sphyxis.github.io/bitburner-src/

Added features:
CharityORG - Recruit your volunteers to do good in the world
Lottery - A fully functioning lottery store with Pick2,3,4, Keeno and Lotto6/49
Full API support for Lottery and the CharityORG
Charitable actions in the slums
Dynamic Events that increase in difficulty as you level your volunteers
Randomized prizes and items
Quantum Tickets - Lottery tickets that regenerate after every install
Charity Servers - Low Ram/High Core personal servers that follow you everywhere

I am looking for feedback, improvements, names for the charitable events in the slums, and anything else that will help make this challenging BN.

Screenshot 2024-01-20 130452
Screenshot 2024-01-20 130616
Screenshot 2024-01-20 130556
Screenshot 2024-01-20 130537

### CODE REVIEW

**Comment by Snarling** [comment]
I tried to rebase this but because it's so many commits, it asked me to address loads of conflicts for all the older commits (that were eventually deconflicted in future commits), so I aborted that rebase attempt. I think this might be due to you using a different method to keep your branch up to date with dev. The rebase method I use requires adding this PR's commits on top of the existing dev branch, so if there are conflicts with earlier commits, it expects me to deconflict those even though they were addressed later on by one of the remote tracking branch merges.

I tried loading up the linked site http://184.70.125.222:8000/ but it was pretty slow to load for me. Might be better to host it on github similar to what @ficocelliguy did for BN14 testing on PR #934 

In initial testing I noticed a discrepancy while trying to start a charity (in BN 14.2 if it matters):
image

The text makes it seem like the self-fund cost is 250m, but then the cost on the button is 50m

**Comment by Snarling** [comment]
Some typos I noticed:

"Spead the word" (member task, should be "Spread")
"Stop a Robery" (slums task, should be "Robbery")
"Patrole the streets" (slums task, should be "Patrol")

File: src/Augmentation/CircadianModulator.ts [hunk 1]
```diff
@@ -110,6 +110,15 @@ function getRandomBonus(): CircadianBonus {
...
     },
+    {
+      bonuses: {
+        charity_success: 2,
+        charity_money: 2,
+      },
+      description:
+        "Increases the player's charity success rate by 100%.\n" +
+        "Increases the amount of money the player gains from charity by 100%.",
+    },
```
Comment (Snarling) [inline src/Augmentation/CircadianModulator.ts]: This augment doesn't have any other effects that are for lategame mechanics, should probably remove this effect.

File: markdown/bitburner.charityaugmentationareas.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+[Home](./index.md) &gt; [bitburner](./bitburner.md) &gt; [CharityAugmentationAreas](./bitburner.charityaugmentationareas.md)
+
+## CharityAugmentationAreas type
+
+Charity Augmentation Areas.
+
+**Signature:**
+
+```typescript
+type CharityAugmentationAreas =
```
Comment (Snarling) [inline markdown/bitburner.charityaugmentationareas.md]: Type should not be pluralized.

e.g. 
```ts
const augArea: CharityAugmentationAreas = "hacking";
```
In usage it is clear that something with this type is referring to a single area, not multiple areas, so the plural type name is confusing.

File: markdown/bitburner.charitybannerpieceinfo.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+interface CharityBannerPieceInfo 
+```
+
+## Properties
+
+|  Property | Modifiers | Type | Description |
+|  --- | --- | --- | --- |
+|  [effects](./bitburner.charitybannerpieceinfo.effects.md) |  | [CharityBannerFragmentInfo](./bitburner.charitybannerfragmentinfo.md)<!-- -->\[\] |  |
+|  [name](./bitburner.charitybannerpieceinfo.name.md) |  | string |  |
+|  [short\_name](./bitburner.charitybannerpieceinfo.short_name.md) |  | string |  |
```
Comment (Snarling) [inline markdown/bitburner.charitybannerpieceinfo.md]: Property name with an underscore is inconsistent with all other naming

File: markdown/bitburner.charityeffectmodifier.area.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<!-- Do not edit this file. It is automatically generated by API Documenter. -->
+
+[Home](./index.md) &gt; [bitburner](./bitburner.md) &gt; [CharityEffectModifier](./bitburner.charityeffectmodifier.md) &gt; [area](./bitburner.charityeffectmodifier.area.md)
+
+## CharityEffectModifier.area property
+
+**Signature:**
+
+```typescript
+area: CharityModifyAreas;
```
Comment (Snarling) [inline markdown/bitburner.charityeffectmodifier.area.md:10]: Type name again should not be pluralized (the area is not an areas)

File: markdown/bitburner.charityeventgendata.deatheffects.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<!-- Do not edit this file. It is automatically generated by API Documenter. -->
+
+[Home](./index.md) &gt; [bitburner](./bitburner.md) &gt; [CharityEventGenData](./bitburner.charityeventgendata.md) &gt; [deathEffects](./bitburner.charityeventgendata.deatheffects.md)
+
+## CharityEventGenData.deathEffects property
+
+**Signature:**
+
+```typescript
+deathEffects: deathEffect[];
```
Comment (Snarling) [inline markdown/bitburner.charityeventgendata.deatheffects.md]: Type name for deathEffect should be capitalized

File: markdown/bitburner.charityeventgendata.modifiers.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<!-- Do not edit this file. It is automatically generated by API Documenter. -->
+
+[Home](./index.md) &gt; [bitburner](./bitburner.md) &gt; [CharityEventGenData](./bitburner.charityeventgendata.md) &gt; [modifiers](./bitburner.charityeventgendata.modifiers.md)
+
+## CharityEventGenData.modifiers property
+
+**Signature:**
+
+```typescript
+modifiers: modifiers[];
```
Comment (Snarling) [inline markdown/bitburner.charityeventgendata.modifiers.md]: `modifiers` should likely be a `Modifier[]` (capitalized and singular), although if Modifier is a type specific to Charity then it should be named something more unique.

**Review by Snarling — COMMENTED** [review]
Started looking at some files, here is a partial review of mostly just API stuff (those files were at the top)

**Comment by Sphyxis** [comment]
Would you like me to try to rebase as well?  I’ve never done that and would need a bit of guidance on how to proceed.

**Comment by Snarling** [comment]
For rebasing I just use Github Desktop. Branch -> Rebase current branch is the way I do it. Because of the way rebasing works (adding all your commits on top of current dev), it'll make you re-resolve every old conflict though. One way to make a rebase easier would be to just squash your previous commits into one before performing the rebase, but that comes at the cost of losing all the individual commits.

I think both ways of keeping the branch up to date are valid, it's just that doing it the way you've been doing it makes it really hard to do the rebase method that I'm more familiar with.

**Comment by d0sboots** [comment]
> I tried to rebase this but because it's so many commits, it asked me to address loads of conflicts for all the older commits (that were eventually deconflicted in future commits), so I aborted that rebase attempt. I think this might be due to you using a different method to keep your branch up to date with dev. The rebase method I use requires adding this PR's commits on top of the existing dev branch, so if there are conflicts with earlier commits, it expects me to deconflict those even though they were addressed later on by one of the remote tracking branch merges.

A common way to deal with this is to squash the commit history first (or squash it into the chunks you want to keep distinct), and *then* do the rebase. Since, as you note, rebase is O(number_of_branch_commits).

Pull+merge continuously (which is the other strategy) is the opposite, being O(number_of_changes_on_head).

Edit: lol, you said the same, I hadn't caught up to the bottom.

**Pull request closed without merging**